nexus/src/app/sagas/instance_start.rs

```diff
@@ -575,18 +580,80 @@ async fn sis_ensure_registered_undo(
         .await
         .map_err(ActionError::action_failed)?;
 
-    osagactx
+    // If the sled successfully unregistered the instance, allow the rest of
+    // saga unwind to restore the instance record to its prior state (without
+    // writing back the state returned from sled agent). Otherwise, try to
+    // reason about the next action from the specific kind of error that was
+    // returned.
+    if let Err(e) = osagactx
         .nexus()
-        .instance_ensure_unregistered(
-            &opctx,
-            &authz_instance,
-            &sled_id,
-            db_instance.runtime(),
-        )
+        .instance_ensure_unregistered(&opctx, &authz_instance, &sled_id)
         .await
-        .map_err(ActionError::action_failed)?;
+    {
+        error!(osagactx.log(),
+               "start saga: failed to unregister instance from sled";
+               "instance_id" => %instance_id,
+               "error" => ?e);
+
+        // If the failure came from talking to sled agent, and the error code
+        // indicates the instance or sled might be unhealthy, manual
+        // intervention is likely to be needed, so try to mark the instance as
+        // Failed and then bail on unwinding.
+        //
+        // If sled agent is in good shape but just doesn't know about the
+        // instance, this saga still owns the instance's state, so allow
+        // unwinding to continue.
+        //
+        // If some other Nexus error occurred, this saga is in bad shape, so
+        // return an error indicating that intervention is needed without trying
+        // to modify the instance further.
+        match e {
+            InstanceStateChangeError::SledAgent(inner)
+                if inner.instance_unhealthy() =>
+            {
+                error!(osagactx.log(),
+                       "start saga: failing instance after unregister failure";
+                       "instance_id" => %instance_id,
+                       "error" => ?inner);
+
+                if let Err(set_failed_error) = osagactx
+                    .nexus()
+                    .mark_instance_failed(
+                        &instance_id,
+                        db_instance.runtime(),
+                        &inner,
+                    )
+                    .await
+                {
+                    error!(osagactx.log(),
+                           "start saga: failed to mark instance as failed";
+                           "instance_id" => %instance_id,
+                           "error" => ?set_failed_error);
+
+                    Err(set_failed_error.into())
+                } else {
+                    Err(inner.0.into())
+                }
+            }
+            InstanceStateChangeError::SledAgent(_) => {
+                info!(osagactx.log(),
+                       "start saga: instance already unregistered from sled";
+                       "instance_id" => %instance_id);
 
-    Ok(())
+                Ok(())
```

Comment:
Oof, yeah, I get this, but it feels a little sketchy to "do nothing" on this pathway. the realm of `!inner.instance_unhealth()` isn't *super* narrow, right?

We don't need to fix this now, I think this is at parity with `main`, but it jumps out to me as a little odd

Reply:
One of the things I like about this change is that it brings out the importance of #3238 and #4226 :) You're right that we could do a lot more to reason carefully about what the different possible error codes mean here.

I added a comment to this match in 560848142.